### DISCUSSION TITLE
Noise and other signal problems

### DISCUSSION THREAD

**Discussion by denkorz** [RomainPastureau/find_delay#3] [Q&A] [answered]
Does it work if there is a noise or pauses in signal?

**Comment by RomainPastureau** [comment, marked as answer]
Hi! Thank you for your interest in this package :)

Regarding the noise, all depends on how noisy the signal is. For example, we tried to use this function to calculate the delay between the recorded onset of an audio clip in our experimental software, and what the subject actually heard.

We know what the subject hears because our magnetoencephalography machine records the audio output in a channel. The signal, however, is very noisy. Despite that, we still manage to find a delay, consistently, of ~200 ms, for all the trials across all participants - without any outlier (actually, there was one, but it was because we tried to correlate the wrong audio file... so it actually worked 100% of the time). 

The only thing is that, depending on the recording, we would get a cross-correlation peak varying between 0.4 and 0.9. But even when the cross-correlation peak was around 0.4, the delay found was the right one.

So, short answer: yes, it will work with noise; however, the more noise, the lower your correlation peak will be. Try with some of your data, and see if 1) it finds the proper delay for one of your files and 2) there is, visually, a clear peak in the cross-correlation. If it works for a few, you should be fine. If you wish, you can also save a graph for each of your delays in a folder, so that you can visually inspect if everything went well.

Regarding pauses, I never tried, but my guess would be no. Because, if you have pauses in the data, then technically the function will return the highest correlation being the moment where there is the longest fragment without a pause. So that would get tricky... Also, there is no way to automatically detect the duration of the pauses in your case, which would probably be what you need.

Let me know if there is anything else!

**Reply by denkorz** [reply]
Thank you a lot for fast and detailed answer!
I've tested my data a bit (most of time was spent to configure Python since this it the last of popular languages I've never used) and it looks the package works perfectly for me :) I used to work with signals correlation in university, but it was Matlab and I never thought I would need to use it later in my life. So thank you a lot!

**Reply by RomainPastureau** [reply]
Awesome! Glad it works for you.

Actually the inception for this function was Matlab! We tried to use their `finddelay` function, but I did not like that they provide the delay value, but not the confidence - hence you can get an accurate result one time and a totally inaccurate one another time, and you wouldn't get a clue. Here, the cross-correlation value is provided, so you can have a better idea if the delay you found is correct or no.

I will close this question, but please do not hesitate to open an issue would you encounter any problems, or if something isn't clear in the documentation.
